UniversalAlgebra/Structures/Graphs.lagda

```diff
@@ -0,0 +1,199 @@
+layout: default
+title : Structures.Graphs
+date : 2021-06-22
+author: William DeMeo
+---
+
+N.B. This module differs from AsRecordsGraphs.lagda in that we assume some universes are lzero (i.e., ℓ₀).
+
+This module implements the graph of a structure.  (See Definition 2 of https://arxiv.org/pdf/2010.04958v2.pdf )
+
+Definition [Graph of a structure]. Let 𝑨 be an (𝑅,𝐹)-structure (relations from 𝑅 and operations from 𝐹).
+The *graph* of 𝑨 is the structure Gr 𝑨 with the same domain as 𝑨 with relations from 𝑅 and together with a (k+1)-ary relation symbol G 𝑓 for each 𝑓 ∈ 𝐹 of arity k, which is interpreted in Gr 𝑨 as all tuples (t , y) ∈ Aᵏ⁺¹ such that 𝑓 t ≡ y.
+
+
+
+\begin{code}
+
+{-# OPTIONS --without-K --exact-split --safe #-}
+
+module Structures.Graphs where
+
+open import Data.Nat using (ℕ)
+open import Data.Fin using (Fin)
+open import Agda.Primitive                        using    ( _⊔_    ;   lsuc     )
+                                                  renaming ( Set    to  Type
+                                                           ; lzero  to ℓ₀        )
+open import Agda.Builtin.Equality                 using    ( _≡_    ;   refl     )
+open import Data.Sum.Base                         using    ( _⊎_                 )
+                                                  renaming ( inj₁   to inl
+                                                           ; inj₂   to inr       )
+open import Data.Product                          using    ( _,_    ;   Σ-syntax
+                                                           ;  Σ     ;   _×_      )
+                                                  renaming ( proj₁  to  fst
+                                                           ; proj₂  to  snd      )
+open import Level                                 using    ( Level  ;  Lift
+                                                           ; lift   ;  lower     )
+open import Function.Base                         using    ( _∘_                 )
+open import Relation.Binary.PropositionalEquality using    ( cong   ; sym
+                                                           ; module ≡-Reasoning  )
+open import Relation.Unary                        using    ( Pred   ; _∈_        )
+
+
+-- Imports from agda-algebras --------------------------------------------------------------
+open import Overture.Preliminaries     using ( ∣_∣ ; _≈_ ; ∥_∥ ; _∙_ ; lower∼lift ; lift∼lower ; 𝟙)
+open import Structures.AsRecordsBasic  using ( signature ; structure ; Sig∅) -- ; Lift-Struc )
+open import Structures.AsRecordsHoms   using ( hom ; 𝒾𝒹 ; ∘-hom ; 𝓁𝒾𝒻𝓉 ; 𝓁ℴ𝓌ℯ𝓇 ; is-hom-rel; is-hom-op)
+open import Relations.Continuous       using ( Rel )
+open import Relations.Extensionality   using ( _≐_ )
+
+open signature
+open structure
+open _⊎_
+
+ℓ₁ : Level
+ℓ₁ = lsuc ℓ₀
+
+Gr-sig : signature → signature → signature
+Gr-sig 𝐹 𝑅 = record { symbol = symbol 𝑅 ⊎ symbol 𝐹 ; arity = arty }
+ where
+ arty : symbol 𝑅 ⊎ symbol 𝐹 → Type ℓ₀
+ arty (inl 𝑟) = (arity 𝑅) 𝑟
```

Comment:
```agda
arty = map (arity 𝑅) (λ f  → arity 𝑅 f ⊎ 𝟙)
```
where `map` is from Data.Product.

Reply:
Also, `⊎ 𝟙` is going to be awkward - why not use `Maybe` instead?

Reply:
Good ideas.  I'll try to implement them.

Reply:
Actually, Maybe is not what I want here.  To construct the purely relational (graph) structure out of a general (realtional+algebraic) structure, I need to convert an n-ary operation symbol f into an (n+1)-ary relation symbol.  Instead, I make the arity type (arity 𝐹) 𝑓 ⊎ 𝟙, which actually gives us more information than merely increasing the arity by 1.

I'm not sure a Maybe type would be appropriate.  I guess you mean the role of the ⊎ 𝟙 would be played by the "nothing" (or "none") constructor of the Maybe...?

Reply:
Also, I think 
`arty = map (arity 𝑅) (λ f  → arity 𝑅 f ⊎ 𝟙)`
is less clear than what I did... but I suppose clarity is in the eye of the beholder to some degree.

Reply:
Yes exactly - those two types are equivalent. The eliminator for `Maybe` is more convenient because it builds-in that the rhs is a unit (so carries no intrinsic information). The only benefit is that the code will be shorter. It may conceptually impeded understanding though, which would argue for having things exactly as you have them.

In any case, I'd be tempted to not go "raw" (i.e. neither `Maybe` nor `⊎ 𝟙` but define a `_+1` operator on types and some patterns to make the manipulation look more semantic than low-level.

Reply:
On clarity: for someone who reads all the code, your version will be more immediately clear, I agree [surprised?]

However, your version then forces any theorem about `arty` to use low-level reasoning that does case-splitting. If you use a higher-level definition, then you can use higher-level theorems about `map`.

So you're buying local clarity for the price of being forced to always do low-level proofs in the future. Unless you rebuild proofs about `arty` that will be equivalent to the ones about `map`.